server/src/main/java/com/flipkart/varadhi/services/VaradhiSubscriptionService.java

```diff
@@ -0,0 +1,112 @@
+package com.flipkart.varadhi.services;
+
+import com.flipkart.varadhi.entities.Project;
+import com.flipkart.varadhi.entities.TopicResource;
+import com.flipkart.varadhi.entities.VaradhiSubscription;
+import com.flipkart.varadhi.exceptions.InvalidOperationForResourceException;
+import com.flipkart.varadhi.spi.db.MetaStore;
+
+import java.util.List;
+import java.util.Objects;
+
+import static com.flipkart.varadhi.entities.VersionedEntity.INITIAL_VERSION;
+
+public class VaradhiSubscriptionService {
+
+    private final MetaStore metaStore;
+
+    public VaradhiSubscriptionService(MetaStore metaStore) {
+        this.metaStore = metaStore;
+    }
+
+    public List<String> getSubscriptionList(String projectName) {
+        return metaStore.getVaradhiSubscriptionNames(projectName);
+    }
+
+    public VaradhiSubscription getSubscription(String subscriptionName, String projectName) {
+        return metaStore.getVaradhiSubscription(subscriptionName, projectName);
+    }
+
+    public VaradhiSubscription createSubscription(VaradhiSubscription subscription) {
+        // check project and topic exists
+        Project project = metaStore.getProject(subscription.getProject());
+        TopicResource topic = metaStore.getTopicResource(subscription.getTopic(), project.getName());
+        if (subscription.isGrouped() && !topic.isGrouped()) {
+            throw new IllegalArgumentException(
+                    "Cannot create grouped Subscription as it's Topic(%s:%s) is not grouped".formatted(
+                            project.getName(), subscription.getTopic()));
+        }
+
+        // check for duplicate subscription
+        if (metaStore.checkVaradhiSubscriptionExists(subscription.getName(), project.getName())) {
+            throw new IllegalArgumentException(
+                    "Subscription(%s:%s) already exists".formatted(project.getName(), subscription.getName()));
+        }
+
+        // ensure version is set correctly
+        subscription.setVersion(INITIAL_VERSION);
+
+        // persist
+        metaStore.createVaradhiSubscription(subscription);
+
+        return subscription;
+    }
+
+    public VaradhiSubscription updateSubscription(VaradhiSubscription update) {
+        String subscriptionName = update.getName();
+        String projectName = update.getProject();
+
+        // check subscription exist
+        if (!metaStore.checkVaradhiSubscriptionExists(subscriptionName, projectName)) {
+            throw new IllegalArgumentException(
```

Comment:
can't we simply rely on `getVaradhiSubscription` method to verify whether it exists or not to avoid multiple calls to zookeeper?

Reply:
done